### PR TITLE
020 generate report

### PR DESCRIPTION
**Pull request by Adnanhoq**

### CODE REVIEW

**Review by AZZohaib — APPROVED** [review]

**Review by AZZohaib — APPROVED** [review]

File: test/unit/services/JobRoleServiceTests.ts [hunk 1]
```diff
@@ -258,5 +263,49 @@ describe('JobRoleService', function () {
       })
 
     })
+    describe('Generate A Report of All Job Roles', function () {
+
+      it('should throw an error when the API returns a 200 status', async () => {
```
Comment (JennicaMistry) [inline test/unit/services/JobRoleServiceTests.ts:268]: Why does this 200 status test throw an error?

File: src/controllers/JobRoleController.ts [hunk 2]
```diff
@@ -1,12 +1,12 @@
 import express from "express"
```
Comment (JennicaMistry) [inline src/controllers/JobRoleController.ts:1]: There are missing tests for JobRoleController.

File: views/report.njk [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{% extends "./layouts/main.njk" %}
```
Comment (JennicaMistry) [inline views/report.njk:1]: Delete this page as it is not used.

File: views/css/jobRolesStyle.css [hunk 4]
```diff
@@ -20,6 +20,17 @@ h1{
     filter: contrast(0.9);
 }
 
+
+@keyframes spinCursor {
+    0% { transform: rotate(0deg); }
+    100% { transform: rotate(360deg); }
+  }
+  
+  .spinning-cursor {
```
Comment (JennicaMistry) [inline views/css/jobRolesStyle.css:29]: We are missing pictures of the UI and how the spinning-cursor looks on the page.

**Review by JennicaMistry — CHANGES_REQUESTED** [review]

File: src/controllers/JobRoleController.ts (hunk 2 above)
Comment (Adnanhoq) [reply src/controllers/JobRoleController.ts:1]: Added

File: src/controllers/JobRoleController.ts [hunk 5]
```diff
@@ -15,18 +15,33 @@ export const getAllJobRoles = async (req: express.Request, res: express.Response
...
-    }
+    } 
 }
 
 export const getSingleJobRole = async (req: express.Request, res: express.Response): Promise<void> => {
     try {
         const jobRole = await getJobRoleById(req.params.id, req.session.token ?? '');
         
-        const jobResponsibilitiesSplit = jobRole.responsibilities.split("•").filter(responsibility => responsibility.trim() !== '');
+        const jobResponsibilitiesSplit = jobRole.responsibilities.split("*").filter(responsibility => responsibility.trim() !== '');
```
[the remark this answers is not in the document]
Comment (Adnanhoq) [reply src/controllers/JobRoleController.ts:25]: I shall let her know

File: src/controllers/JobRoleController.ts [hunk 6]
```diff
@@ -15,18 +15,33 @@ export const getAllJobRoles = async (req: express.Request, res: express.Response
...
         res.locals.errormessage = (e as Error).message;
         res.render('errorPage.njk', {error: e as Error, token: req.session.token});
     }
 }
+
+export const downloadJobRolesReport = async (req: express.Request, res: express.Response): Promise<void> => {
+    try {
+        const jobRolesReport = await getReportOfJobRoles();
+
+        // Send the CSV data as a downloadable response
```
[the remark this answers is not in the document]
Comment (Adnanhoq) [reply src/controllers/JobRoleController.ts:38]: Removed

File: src/controllers/JobRoleController.ts [hunk 7]
```diff
@@ -15,18 +15,33 @@ export const getAllJobRoles = async (req: express.Request, res: express.Response
...
+        res.setHeader('Content-Disposition', 'attachment; filename=job_roles_report.csv');
+        res.setHeader('Content-Type', 'text/csv');
+        res.send(jobRolesReport);
+        
+    } catch (e) {
+        console.error('Error generating report:', e);
```
[the remark this answers is not in the document]
Comment (Adnanhoq) [reply src/controllers/JobRoleController.ts:44]: 'Failed to generate report, Please try again'

File: test/unit/services/JobRoleServiceTests.ts (hunk 1 above)
Comment (Adnanhoq) [reply test/unit/services/JobRoleServiceTests.ts:268]: Fixed

File: views/css/jobRolesStyle.css (hunk 4 above)
Comment (Adnanhoq) [reply views/css/jobRolesStyle.css:29]: Added to the ticket

File: views/report.njk (hunk 3 above)
Comment (Adnanhoq) [reply views/report.njk:1]: Deleted

**Review by Leebee1392 — DISMISSED** [review]

**Review by AZZohaib — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Review by AZZohaib — APPROVED** [review]

**Pull request merged**